### PR TITLE
Warehouse management

### PR DESCRIPTION
**Pull request by Mychalevytsh**

### CODE REVIEW

**Comment by Mychalevytsh** [comment]
service

File: src/main/java/com/ita/if103java/ims/dao/impl/AbstractDao.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        LOG.error(message, e);
+        return e;
+    }
+
+    protected CRUDException crudException(RuntimeException e, String message) {
+        CRUDException exception = new CRUDException(message, e);
+        LOG.error(message, e);
+        return exception;
+    }
+}
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/dao/impl/AbstractDao.java]: Not sure, but looks like that not needed anymore

File: src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java [hunk 2]
```diff
@@ -56,18 +57,19 @@ private PreparedStatement createWarehouseStatement(Warehouse warehouse, Connecti
             return jdbcTemplate.query(Queries.SQL_SELECT_ALL_WAREHOUSES, warehouseRowMapper);
 
         } catch (DataAccessException e) {
-            throw crudException(e.getMessage(), "get", "*");
+            throw new EntityNotFoundException("Error during finding all warehouses", e);
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java]:
```java
new WarehouseNotFoundException(....)
```

File: src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java [hunk 3]
```diff
@@ -99,55 +101,71 @@ public Warehouse update(Warehouse warehouse) {
...
         int status;
         try {
-            status = jdbcTemplate.update(Queries.SQL_DELETE_WAREHOUSE, false, id);
+            status = jdbcTemplate.update(Queries.SQL_SET_ACTIVE_STATUS_WAREHOUSE, false, id);
         } catch (DataAccessException e) {
-            throw crudException(e.getMessage(), "delete", "id = " + id);
+            throw new CRUDException("Error during soft `delete` {id = " + id + "}", e);
+        }
+        if (status == 0) {
+            throw new WarehouseNotFoundException("Failed to obtain user during soft `delete` {id = " + id + "}");
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java]: to obtain `warehouse`

File: src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java [hunk 4]
```diff
@@ -99,55 +101,71 @@ public Warehouse update(Warehouse warehouse) {
...
-            "capacity = ?, is_bottom = ?, top_warehouse_id = ?" + "active = ? WHERE id = ?";
-        static final String SQL_DELETE_WAREHOUSE = "DELETE FROM warehouses WHERE id = ?";
+        static final String SQL_UPDATE_WAREHOUSE = "UPDATE warehouses SET name= ?, info = ?," +
+            "capacity = ?, is_bottom = ?, top_warehouse_id = ?, active = ? WHERE id = ?";
+
+        static final String SQL_DELETE_WAREHOUSE_BY_ID = "DELETE FROM warehouses WHERE id = ?";
+
+        static final String SQL_SELECT_CHILDREN_BY_ID = "SELECT * FROM warehouses WHERE top_warehouse_id = ?";
+
+        static final String SQL_SET_ACTIVE_STATUS_WAREHOUSE = "UPDATE users SET active = ? WHERE id = ?";
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java]: Maybe here should be
```java
SQL_SET_ACTIVE_STATUS_WAREHOUSE = "UPDATE warehouses SET active = ? WHERE id = ?";
```

File: src/main/java/com/ita/if103java/ims/dto/WarehouseDto.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+@Component
+public class WarehouseDto implements Serializable {
+
+    @NotNull(message = "This field must be filled with the auto-generator during creating new warehouse")
+    private long id;
+    @NotBlank(message = "This field mustn't be empty. Please, enter name of warehouse")
+    private String name;
+    @NotBlank(message = "This field mustn't be empty. Please, enter some information about warehouse")
+    private String info;
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/dto/WarehouseDto.java]: Maybe at least `info` can be blank

File: src/main/java/com/ita/if103java/ims/mapper/WarehouseDtoMapper.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import com.ita.if103java.ims.entity.Warehouse;
+import org.springframework.stereotype.Component;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Component
+public class WarehouseDtoMapper {
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/mapper/WarehouseDtoMapper.java]: extends from AbstractEntityDtoMapper<Warehouse, WarehouseDto>

File: src/main/java/com/ita/if103java/ims/mapper/WarehouseDtoMapper.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            .collect(Collectors.toList());
+    }
+
+    public List<WarehouseDto> convertToWarehouseDtoList(List<Warehouse> warehouseList) {
+        return Optional.ofNullable(warehouseList)
+            .orElse(Collections.emptyList())
+            .stream()
+            .map(this::convertWarehouseToWarehouseDto)
+            .collect(Collectors.toList());
+    }
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/mapper/WarehouseDtoMapper.java]: And then get rid off those methods :)

File: src/main/java/com/ita/if103java/ims/service/impl/WarehouseServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    public List<WarehouseDto> findWarehousesByParentId(Long parentId) {
+        return warehouseDtoMapper.convertToWarehouseDtoList(warehouseDao.findChildrenByID(parentId));
+    }
+
+    @Override
+    public WarehouseDto update(WarehouseDto warehouseDto) {
+        Warehouse updatedWarehouse = warehouseDtoMapper.convertWarehouseDtoToWarehouse(warehouseDto);
+        //Activating status can't be changed in this way
+        Warehouse DBWarehouse = warehouseDao.findById(updatedWarehouse.getId());
```
Comment (rostIvan) [inline src/main/java/com/ita/if103java/ims/service/impl/WarehouseServiceImpl.java]: DBWarehouse to CamelCase
```java
Warehouse dbWarehouse = warehouseDao.findById(updatedWarehouse.getId());
// or
Warehouse warehouse = warehouseDao.findById(updatedWarehouse.getId());
```

File: src/main/java/com/ita/if103java/ims/dto/WarehouseDto.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Null;
+import java.io.Serializable;
+import java.util.Objects;
+
+@Component
+public class WarehouseDto implements Serializable {
+    private long id;
+    private String name;
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dto/WarehouseDto.java]: Name should be required, accountId too

File: src/main/java/com/ita/if103java/ims/exception/CRUDException.java [hunk 10]
```diff
@@ -17,4 +17,4 @@ public CRUDException(Throwable cause) {
         super(cause);
     }
 
-}
\ No newline at end of file
+}
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/exception/CRUDException.java]: Not needed change, please revert it back

File: src/main/java/com/ita/if103java/ims/exception/EntityNotFoundException.java [hunk 11]
```diff
@@ -16,5 +17,4 @@ public EntityNotFoundException(String message, Throwable cause) {
     public EntityNotFoundException(Throwable cause) {
         super(cause);
     }
-
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/exception/EntityNotFoundException.java]: Not needed change, please revert it back

File: src/main/java/com/ita/if103java/ims/mapper/WarehouseDtoMapper.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            warehouseDto.setCapacity(warehouse.getCapacity());
+            warehouseDto.setBottom(warehouse.isBottom());
+            warehouseDto.setAccountID(warehouse.getAccountID());
+            warehouseDto.setParentID(warehouse.getParentID());
+            warehouseDto.setActive(warehouse.isActive());
+            warehouseDto.setTopWarehouseID(warehouse.getTopWarehouseID());
+            return warehouseDto;
+        }
+
+    }
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/mapper/WarehouseDtoMapper.java]: Please move body of these methods to appropriate methods below. and then remove these methods

File: src/main/java/com/ita/if103java/ims/service/impl/WarehouseServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public WarehouseDto findWarehouseById(Long id) {
+        return warehouseDtoMapper.convertWarehouseToWarehouseDto(warehouseDao.findById(id));
+    }
+
+    @Override
+    public List<WarehouseDto> findWarehousesByParentId(Long parentId) {
+        return warehouseDtoMapper.toDtoList(warehouseDao.findChildrenByID(parentId));
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/service/impl/WarehouseServiceImpl.java]: I believe we talk about this and methods names should reflect what they do. In this case findWarehousesByTopLevelId, isn't it?

File: pom.xml [hunk 14]
```diff
@@ -96,6 +96,14 @@
...
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>13</source>
+                    <target>13</target>
+                </configuration>
+            </plugin>
```
Comment (ikrasWork) [inline pom.xml]: duplicate of the plugin, please check rows above. Please remove this redundancy.

**Pull request merged**